enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+administrator has to add all subdomains in the EgressFirewall rules. This becomes difficult when
+subdomains are dynamically added/removed as each one has to be added/removed individually from the
+EgressFirewall rules. Currently, wildcard DNS names are not supported in EgressFirewall. However,
+even if the support is added to EgressFirewall, the integration between CoreDNS and EgressFirewall
+needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.
+
+
+### User Stories
+
+* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow/deny
+access to them even if the IPs associated with the corresponding DNS records change dynamically.
+* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
+allow/deny access to all the subdomains belonging to the wildcard DNS names.
+* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
+in EgressFirewall rules can be tracked.
+* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
+rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
+* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the
+new plugin enabled.
+
+### Goals
+
+* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall rule containing
+a DNS name if the IPs associated with it change dynamically.
+* Support usage of wildcard DNS names in EgressFirewall rules.
+* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current
+IPs and the corresponding TTL information.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin.
+* Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
+(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
+for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
+be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
+based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
+CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an
+OpenShift cluster adminisrator.
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the
```

Comment:
The new plugin should only be used by EgressFirewall, not any other lookup mechanisms.

Reply:
The DNS query will always be processed by the new plugin as per the processing logic of CoreDNS plugin chain (https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/). Thus, the matching of the DNS name in the query needs to be done with the `DNSName` CRs and if it matches only then the actual logic of the new plugin will be executed.

Reply:
> * What is the impact if there are no DNSName CRs?  Is it only the time it takes to execute `if len(dn.Items) == 0 { /* Execute the next plugin. */}`?

That's correct.

> * What is the impact if there are DNSName CRs but none that matches the query?  Is it the cost of a map lookup?  Is the cost greater when rules use wildcards?  I'm curious what data structure the plugin will use for looking up the DNSName CR for a query.

We can have 2 separate maps for storing regular and wildcard DNS names. The check should happen for both types of DNS names. For wildcard DNS name, the check should happen for each subdomain as well. For example, if a DNS lookup is done for `sub1.sub2.example.com`, then the lookup needs to happen for `*.sub2.example.com` and `*.example.com`. So for wildcards the cost will be higher.

> * What is the impact if some DNSName CR matches the query?  Presumably the status update can happen asynchronously, but does the plugin force a delay while OVN-K updates?

In the current version of the EP, there is no delay that is enforced while waiting for OVN-K updates. However, it has been suggested in this https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162 to wait for OVN-K updates before responding to DNS queries.

Reply:
> For wildcard DNS name, the check should happen for each subdomain as well. For example, if a DNS lookup is done for `sub1.sub2.example.com`, then the lookup needs to happen for `*.sub2.example.com` and `*.example.com`. So for wildcards the cost will be higher.

In https://github.com/openshift/enhancements/pull/1335#discussion_r1157154525, you said that a `*` would match only one label, so I would expect that for a query for `sub1.sub2.example.com`, the plugin would look up `sub1.sub2.example.com` in the regular map and `*.sub2.example.com` (or simply `sub2.example.com`) in the wildcard map.  Is that what you meant, or am I misunderstanding something?

Reply:
Sorry. My bad. It should only try to match `*.sub2.example.com` for wildcard as `*` will only match one label. Thanks for pointing that out.

Reply:
@arkadeepsen following up on this discussion, is there any conclusion/update to Miciah's question on :

> What is the impact if some DNSName CR matches the query? Presumably the status update can happen asynchronously, but does the plugin force a delay while OVN-K updates?

It seems like this might need update.  If possible, just answer the question here instead of reference to another part of the doc.

Reply:
In a meeting with @danwinship, @Miciah and @npinaeva, it was decided that the CoreDNS plugin will not wait for the OVN-K updates. So there won't be any forced delay on the plugin from the OVN-K side.